securedrop_client/logic.py

```diff
@@ -574,57 +575,114 @@ def on_reply_download_failure(self, exception: Exception) -> None:
+            return
+
+        path_to_file_with_original_name = os.path.join(self.data_dir, file.original_filename)
+        if not os.path.exists(path_to_file_with_original_name):
+            os.link(filepath, path_to_file_with_original_name)
+
+        if not self.qubes:
+            return
+
+        command = "qvm-open-in-vm"
+        args = ['$dispvm:sd-svs-disp', path_to_file_with_original_name]
+        process = QProcess(self)
+        process.start(command, args)
+
+    def run_export_preflight_checks(self, file_uuid: str):
         '''
-        Run preflight checks to make sure the Export VM is configured correctly
+        Run preflight checks to make sure the Export VM is configured correctly. Also check if the
+        file we're about to export exists before asking the user to connect their thumb drive and
+        type in their passphrase to unlock it.
         '''
-        logger.debug('Calling export preflight checks from thread {}'.format(
-            threading.current_thread().ident))
+        logger.info('Running export preflight checks')
+
+        file = self.get_file(file_uuid)
+        fn_no_ext, dummy = os.path.splitext(os.path.splitext(file.filename)[0])
+        filepath = os.path.join(self.data_dir, fn_no_ext)
+        if not os.path.exists(filepath):
+            msg = _('Could not export {}. File does not exist.'.format(file.original_filename))
+            storage.mark_as_not_downloaded(file_uuid, self.session)
+            self.sync_api()
+            logger.debug(msg)
+            self.gui.update_error_status(msg)
+            return
 
         if not self.qubes:
             return
 
         self.export.begin_preflight_check.emit()
 
     def export_file_to_usb_drive(self, file_uuid: str, passphrase: str) -> None:
+        '''
+        Send the file specified by file_uuid to the Export VM with the user-provided passphrase for
+        unlocking the attached transfer device.
+
+        Once a file is downloaded, it exists in the data directory with the same filename as the
+        server, except with the .gz.gpg stripped off. In order for the user to know which
+        application to open the file in, we export the file with a different name: the original
+        filename which includes the file extesion.
+
+        If the file is missing, update the db so that is_downloaded is set to False.
+        '''
         file = self.get_file(file_uuid)
+        logger.info('Exporting file {}'.format(file.original_filename))
+
+        fn_no_ext, dummy = os.path.splitext(os.path.splitext(file.filename)[0])
+        filepath = os.path.join(self.data_dir, fn_no_ext)
+        if not os.path.exists(filepath):
+            msg = _('Could not export {}. File does not exist.'.format(file.original_filename))
+            storage.mark_as_not_downloaded(file_uuid, self.session)
+            self.sync_api()
+            logger.debug(msg)
+            self.gui.update_error_status(msg)
```

Comment:
the other thing we need to decide is how to close the export dialog in the file not found case. some options:

1. Add a little method on the `Export` object that closes the `ExportDialog`, which we can call from the controller. This _would_ work, but I don't love this approach because there already is a method for handling errors in the export process which is via the pre_flight and usb failure signals.
2. At this failure stage, emit the pre_flight or usb export failure signal and modify `ExportDialog._update` to handle a new error status: when the file is not there. The `ExportDialog` can just close in this case because we are showing the error in the main window. 
3. Move this file not found logic to _before_ the export dialog even pops up. 

I'm a fan of either 2 or 3. What do you think?

Reply:
yeah, right now, we just wait for the gui to refresh before the dialog closes when the file is missing. i think using the failure signal makes the most sense since the file _could_ be deleted while the dialog is open so it would be good to be able to have code in place to close the dialog upon a file missing error.